From 62c58f011e7069aa45e980930cdfa2d937920a77 Mon Sep 17 00:00:00 2001
From: Hans de Goede <hdegoede@redhat.com>
Date: Fri, 8 Jun 2012 10:57:54 +0200
Subject: [PATCH 1/4] gstcamerabin: Fix spelling error in debug logging

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 gst/camerabin/gstcamerabin.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gst/camerabin/gstcamerabin.c b/gst/camerabin/gstcamerabin.c
index a17e02c..8a36970 100644
--- a/gst/camerabin/gstcamerabin.c
+++ b/gst/camerabin/gstcamerabin.c
@@ -1517,7 +1517,7 @@ gst_camerabin_set_capsfilter_caps (GstCameraBin * camera, GstCaps * new_caps)
   if (camera->src_zoom_filter)
     g_object_set (G_OBJECT (camera->src_zoom_filter), "caps", new_caps, NULL);
   gst_camerabin_update_aspect_filter (camera, new_caps);
-  GST_INFO_OBJECT (camera, "udpated");
+  GST_INFO_OBJECT (camera, "updated");
 }
 
 /*
-- 
1.7.10.2

