diff -up gst-plugins-bad-0.10.13/ext/assrender/gstassrender.c.new-ass-api gst-plugins-bad-0.10.13/ext/assrender/gstassrender.c
--- gst-plugins-bad-0.10.13/ext/assrender/gstassrender.c.new-ass-api	2009-03-06 18:37:51.000000000 +0100
+++ gst-plugins-bad-0.10.13/ext/assrender/gstassrender.c	2009-08-10 13:08:56.000000000 +0200
@@ -297,13 +297,12 @@ gst_assrender_setcaps_video (GstPad * pa
       gst_structure_get_int (structure, "height", &render->height)) {
     ret = gst_pad_set_caps (render->srcpad, caps);
     ass_set_frame_size (render->ass_renderer, render->width, render->height);
-    /* FIXME: Does this expect aspect ratio or pixel aspect ratio? */
-    ass_set_aspect_ratio (render->ass_renderer,
+    ass_set_aspect_ratio (render->ass_renderer, 
+        ((gdouble) (render->width)) / ((gdouble) (render->height)),
         ((gdouble) par_n) / ((gdouble) par_d));
     ass_set_font_scale (render->ass_renderer, 1.0);
     ass_set_hinting (render->ass_renderer, ASS_HINTING_NATIVE);
-    ass_set_fonts (render->ass_renderer, "Arial", "sans-serif");
-    ass_set_fonts (render->ass_renderer, NULL, "Sans");
+    ass_set_fonts (render->ass_renderer, NULL, "Sans", 1, NULL, 1);
     ass_set_margins (render->ass_renderer, 0, 0, 0, 0);
     ass_set_use_margins (render->ass_renderer, 0);
 
@@ -383,7 +382,7 @@ gst_assrender_chain_video (GstPad * pad,
   gint64 start, stop, clip_start = 0, clip_stop = 0;
   double timestamp;
   double step;
-  ass_image_t *ass_image;
+  ASS_Image *ass_image;
 
   render = GST_ASSRENDER (GST_PAD_PARENT (pad));
 
diff -up gst-plugins-bad-0.10.13/ext/assrender/gstassrender.h.new-ass-api gst-plugins-bad-0.10.13/ext/assrender/gstassrender.h
--- gst-plugins-bad-0.10.13/ext/assrender/gstassrender.h.new-ass-api	2009-03-06 18:37:51.000000000 +0100
+++ gst-plugins-bad-0.10.13/ext/assrender/gstassrender.h	2009-08-10 12:56:09.000000000 +0200
@@ -44,9 +44,9 @@ struct _Gstassrender
 
   gint width, height;
 
-  ass_library_t *ass_library;
-  ass_renderer_t *ass_renderer;
-  ass_track_t *ass_track;
+  ASS_Library *ass_library;
+  ASS_Renderer *ass_renderer;
+  ASS_Track *ass_track;
 
   gboolean renderer_init_ok, track_init_ok, enable, embeddedfonts;
 };
