diff -up gst-plugins-bad-0.10.16/ext/celt/gstceltdec.c~ gst-plugins-bad-0.10.16/ext/celt/gstceltdec.c
--- gst-plugins-bad-0.10.16/ext/celt/gstceltdec.c~	2009-10-12 12:23:38.000000000 +0200
+++ gst-plugins-bad-0.10.16/ext/celt/gstceltdec.c	2009-11-07 08:01:54.000000000 +0100
@@ -498,13 +498,12 @@ celt_dec_chain_parse_header (GstCeltDec 
     goto invalid_header;
 
   dec->mode =
-      celt_mode_create (dec->header.sample_rate, dec->header.nb_channels,
-      dec->header.frame_size, &error);
+      celt_mode_create (dec->header.sample_rate, dec->header.frame_size, &error);
   if (!dec->mode)
     goto mode_init_failed;
 
   /* initialize the decoder */
-  dec->state = celt_decoder_create (dec->mode);
+  dec->state = celt_decoder_create (dec->mode, dec->header.nb_channels, &error);
   if (!dec->state)
     goto init_failed;
 
@@ -543,7 +542,7 @@ mode_init_failed:
 init_failed:
   {
     GST_ELEMENT_ERROR (GST_ELEMENT (dec), STREAM, DECODE,
-        (NULL), ("couldn't initialize decoder"));
+        (NULL), ("couldn't initialize decoder: %d", error));
     return GST_FLOW_ERROR;
   }
 nego_failed:
diff -up gst-plugins-bad-0.10.16/ext/celt/gstceltenc.c~ gst-plugins-bad-0.10.16/ext/celt/gstceltenc.c
--- gst-plugins-bad-0.10.16/ext/celt/gstceltenc.c~	2009-10-12 12:23:38.000000000 +0200
+++ gst-plugins-bad-0.10.16/ext/celt/gstceltenc.c	2009-11-07 08:04:19.000000000 +0100
@@ -577,16 +577,16 @@ gst_celt_enc_setup (GstCeltEnc * enc)
   enc->setup = FALSE;
 
   enc->mode =
-      celt_mode_create (enc->rate, enc->channels, enc->frame_size, &error);
+      celt_mode_create (enc->rate, enc->frame_size, &error);
   if (!enc->mode)
     goto mode_initialization_failed;
 
   celt_mode_info (enc->mode, CELT_GET_FRAME_SIZE, &enc->frame_size);
 
-  celt_header_init (&enc->header, enc->mode);
+  celt_header_init (&enc->header, enc->mode, enc->channels);
   enc->header.nb_channels = enc->channels;
 
-  enc->state = celt_encoder_create (enc->mode);
+  enc->state = celt_encoder_create (enc->mode, enc->channels, &error);
   if (!enc->state)
     goto encoder_creation_failed;
 
@@ -609,7 +609,7 @@ mode_initialization_failed:
   return FALSE;
 
 encoder_creation_failed:
-  GST_ERROR_OBJECT (enc, "Encoder creation failed");
+  GST_ERROR_OBJECT (enc, "Encoder creation failed: %d", error);
   return FALSE;
 }
 
