diff -uNr gst-plugins-bad-0.10.23/ext/timidity/gsttimidity.c gst-plugins-bad-0.10.23.timidity/ext/timidity/gsttimidity.c
--- gst-plugins-bad-0.10.23/ext/timidity/gsttimidity.c	2011-12-30 19:24:28.000000000 +0100
+++ gst-plugins-bad-0.10.23.timidity/ext/timidity/gsttimidity.c	2016-10-06 10:37:54.076801395 +0200
@@ -615,7 +615,7 @@
     GST_DEBUG_OBJECT (timidity, "Parsing song");
 
     stream =
-        mid_istream_open_mem (timidity->mididata, timidity->mididata_size, 0);
+        mid_istream_open_mem (timidity->mididata, timidity->mididata_size);
 
     timidity->song = mid_song_load (stream, timidity->song_options);
     mid_istream_close (stream);
