diff -up gst-plugins-bad-0.10.10/gst/modplug/Makefile.am.foo gst-plugins-bad-0.10.10/gst/modplug/Makefile.am
--- gst-plugins-bad-0.10.10/gst/modplug/Makefile.am.foo	2009-01-06 11:56:05.000000000 +0100
+++ gst-plugins-bad-0.10.10/gst/modplug/Makefile.am	2009-01-21 10:01:03.000000000 +0100
@@ -1,10 +1,8 @@
-SUBDIRS=libmodplug .
-
 plugin_LTLIBRARIES = libgstmodplug.la
 
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 libgstmodplug_la_LIBTOOLFLAGS = --tag=disable-static
 
diff -up gst-plugins-bad-0.10.10/gst/modplug/Makefile.in.foo gst-plugins-bad-0.10.10/gst/modplug/Makefile.in
--- gst-plugins-bad-0.10.10/gst/modplug/Makefile.in.foo	2009-01-19 23:06:17.000000000 +0100
+++ gst-plugins-bad-0.10.10/gst/modplug/Makefile.in	2009-01-21 10:02:33.000000000 +0100
@@ -85,7 +85,6 @@ pluginLTLIBRARIES_INSTALL = $(INSTALL)
 LTLIBRARIES = $(plugin_LTLIBRARIES)
 am__DEPENDENCIES_1 =
 libgstmodplug_la_DEPENDENCIES =  \
-	$(top_builddir)/gst/modplug/libmodplug/libmodplug.la \
 	$(am__DEPENDENCIES_1)
 am_libgstmodplug_la_OBJECTS = libgstmodplug_la-gstmodplug.lo
 libgstmodplug_la_OBJECTS = $(am_libgstmodplug_la_OBJECTS)
@@ -436,11 +435,11 @@ sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = libmodplug .
+SUBDIRS = .
 plugin_LTLIBRARIES = libgstmodplug.la
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 libgstmodplug_la_LIBTOOLFLAGS = --tag=disable-static
 noinst_HEADERS = gstmodplug.h
diff -up gst-plugins-bad-0.10.10/gst/modplug/gstmodplug.cc.foo gst-plugins-bad-0.10.10/gst/modplug/gstmodplug.cc
--- gst-plugins-bad-0.10.10/gst/modplug/gstmodplug.cc.foo	2009-01-06 11:56:05.000000000 +0100
+++ gst-plugins-bad-0.10.10/gst/modplug/gstmodplug.cc	2009-01-21 10:01:03.000000000 +0100
@@ -43,8 +43,8 @@
 #include "config.h"
 #endif
 
-#include "libmodplug/stdafx.h"
-#include "libmodplug/sndfile.h"
+#include <libmodplug/stdafx.h>
+#include <libmodplug/sndfile.h>
 
 #include "gstmodplug.h"
 
