diff -up gst-plugins-bad-0.10.8/gst/modplug/Makefile.am.foo gst-plugins-bad-0.10.8/gst/modplug/Makefile.am
--- gst-plugins-bad-0.10.8/gst/modplug/Makefile.am.foo	2008-04-20 15:06:06.000000000 +0200
+++ gst-plugins-bad-0.10.8/gst/modplug/Makefile.am	2008-08-01 15:58:29.000000000 +0200
@@ -1,10 +1,8 @@
-SUBDIRS=libmodplug .
-
 plugin_LTLIBRARIES = libgstmodplug.la
 
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 
 noinst_HEADERS = gstmodplug.h
diff -up gst-plugins-bad-0.10.8/gst/modplug/Makefile.in.foo gst-plugins-bad-0.10.8/gst/modplug/Makefile.in
--- gst-plugins-bad-0.10.8/gst/modplug/Makefile.in.foo	2008-07-31 23:54:41.000000000 +0200
+++ gst-plugins-bad-0.10.8/gst/modplug/Makefile.in	2008-08-01 16:00:44.000000000 +0200
@@ -84,7 +84,6 @@ pluginLTLIBRARIES_INSTALL = $(INSTALL)
 LTLIBRARIES = $(plugin_LTLIBRARIES)
 am__DEPENDENCIES_1 =
 libgstmodplug_la_DEPENDENCIES =  \
-	$(top_builddir)/gst/modplug/libmodplug/libmodplug.la \
 	$(am__DEPENDENCIES_1)
 am_libgstmodplug_la_OBJECTS = libgstmodplug_la-gstmodplug.lo
 libgstmodplug_la_OBJECTS = $(am_libgstmodplug_la_OBJECTS)
@@ -617,11 +616,11 @@ sbindir = @sbindir@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-SUBDIRS = libmodplug .
+SUBDIRS = .
 plugin_LTLIBRARIES = libgstmodplug.la
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 noinst_HEADERS = gstmodplug.h
 all: all-recursive
diff -up gst-plugins-bad-0.10.8/gst/modplug/gstmodplug.cc.foo gst-plugins-bad-0.10.8/gst/modplug/gstmodplug.cc
--- gst-plugins-bad-0.10.8/gst/modplug/gstmodplug.cc.foo	2008-07-19 15:18:20.000000000 +0200
+++ gst-plugins-bad-0.10.8/gst/modplug/gstmodplug.cc	2008-08-01 15:58:29.000000000 +0200
@@ -43,8 +43,8 @@
 #include "config.h"
 #endif
 
-#include "libmodplug/stdafx.h"
-#include "libmodplug/sndfile.h"
+#include <libmodplug/stdafx.h>
+#include <libmodplug/sndfile.h>
 
 #include "gstmodplug.h"
 
