diff -up gst-plugins-bad-0.10.23/gst/camerabin/Makefile.in.make43 gst-plugins-bad-0.10.23/gst/camerabin/Makefile.in
--- gst-plugins-bad-0.10.23/gst/camerabin/Makefile.in.make43	2012-02-28 20:14:08.000000000 +0100
+++ gst-plugins-bad-0.10.23/gst/camerabin/Makefile.in	2020-04-24 21:46:48.947720695 +0200
@@ -571,7 +571,7 @@ top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 glib_gen_prefix = __gst_camerabin
 glib_gen_basename = gstcamerabin
-enum_headers = $(foreach h,$(glib_enum_headers),\n\#include \"$(h)\")
+enum_headers = $(foreach h,$(glib_enum_headers),\n#include \"$(h)\")
 built_sources = gstcamerabin-marshal.c
 built_headers = gstcamerabin-marshal.h
 BUILT_SOURCES = $(built_sources) $(built_headers)
diff -up gst-plugins-bad-0.10.23/gst/debugutils/Makefile.in.make43 gst-plugins-bad-0.10.23/gst/debugutils/Makefile.in
--- gst-plugins-bad-0.10.23/gst/debugutils/Makefile.in.make43	2012-02-28 20:14:09.000000000 +0100
+++ gst-plugins-bad-0.10.23/gst/debugutils/Makefile.in	2020-04-24 21:46:44.144746539 +0200
@@ -570,7 +570,7 @@ top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 glib_gen_prefix = __gst_debugutils
 glib_gen_basename = debugutils
-enum_headers = $(foreach h,$(glib_enum_headers),\n\#include \"$(h)\")
+enum_headers = $(foreach h,$(glib_enum_headers),\n#include \"$(h)\")
 built_sources = debugutils-marshal.c
 built_headers = debugutils-marshal.h
 BUILT_SOURCES = $(built_sources) $(built_headers)
diff -up gst-plugins-bad-0.10.23/gst-libs/gst/interfaces/Makefile.in.make43 gst-plugins-bad-0.10.23/gst-libs/gst/interfaces/Makefile.in
--- gst-plugins-bad-0.10.23/gst-libs/gst/interfaces/Makefile.in.make43	2012-02-28 20:14:06.000000000 +0100
+++ gst-plugins-bad-0.10.23/gst-libs/gst/interfaces/Makefile.in	2020-04-24 21:46:54.678689858 +0200
@@ -611,7 +611,7 @@ BUILT_SOURCES = \
 	$(built_headers)
 
 CLEANFILES = $(BUILT_SOURCES)
-enum_headers = $(foreach h,$(glib_enum_headers),\n\#include \"$(h)\")
+enum_headers = $(foreach h,$(glib_enum_headers),\n#include \"$(h)\")
 all: $(BUILT_SOURCES)
 	$(MAKE) $(AM_MAKEFLAGS) all-am
 
diff -up gst-plugins-bad-0.10.23/sys/shm/Makefile.in.make43 gst-plugins-bad-0.10.23/sys/shm/Makefile.in
--- gst-plugins-bad-0.10.23/sys/shm/Makefile.in.make43	2012-02-28 20:14:19.000000000 +0100
+++ gst-plugins-bad-0.10.23/sys/shm/Makefile.in	2020-04-24 21:46:38.957774449 +0200
@@ -556,7 +556,7 @@ top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 glib_enum_prefix = gst_shm
-enum_headers = $(foreach h,$(glib_enum_headers),\n\#include \"$(h)\")
+enum_headers = $(foreach h,$(glib_enum_headers),\n#include \"$(h)\")
 plugin_LTLIBRARIES = libgstshm.la
 libgstshm_la_SOURCES = shmpipe.c shmalloc.c gstshm.c gstshmsrc.c gstshmsink.c
 libgstshm_la_CFLAGS = $(GST_CFLAGS) -DSHM_PIPE_USE_GLIB
