diff -up gst-plugins-bad-0.10.7/gst/modplug/Makefile.am.sys-modplug gst-plugins-bad-0.10.7/gst/modplug/Makefile.am
--- gst-plugins-bad-0.10.7/gst/modplug/Makefile.am.sys-modplug	2008-02-23 16:01:32.000000000 +0100
+++ gst-plugins-bad-0.10.7/gst/modplug/Makefile.am	2008-04-24 13:12:34.000000000 +0200
@@ -1,10 +1,8 @@
-SUBDIRS=libmodplug .
-
 plugin_LTLIBRARIES = libgstmodplug.la
 
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 
 noinst_HEADERS = gstmodplug.h
diff -up gst-plugins-bad-0.10.7/gst/modplug/gstmodplug.cc.sys-modplug gst-plugins-bad-0.10.7/gst/modplug/gstmodplug.cc
--- gst-plugins-bad-0.10.7/gst/modplug/gstmodplug.cc.sys-modplug	2007-10-10 13:44:01.000000000 +0200
+++ gst-plugins-bad-0.10.7/gst/modplug/gstmodplug.cc	2008-04-24 13:10:39.000000000 +0200
@@ -44,8 +44,8 @@
 #include "config.h"
 #endif
 
-#include "libmodplug/stdafx.h"
-#include "libmodplug/sndfile.h"
+#include <libmodplug/stdafx.h>
+#include <libmodplug/sndfile.h>
 
 #include "gstmodplug.h"
 
diff -up gst-plugins-bad-0.10.7/gst/modplug/Makefile.in.sys-modplug gst-plugins-bad-0.10.7/gst/modplug/Makefile.in
--- gst-plugins-bad-0.10.7/gst/modplug/Makefile.in.sys-modplug	2008-04-24 01:33:17.000000000 +0200
+++ gst-plugins-bad-0.10.7/gst/modplug/Makefile.in	2008-04-24 13:12:13.000000000 +0200
@@ -530,13 +530,13 @@ sbindir = @sbindir@
 sharedstatedir = @sharedstatedir@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
-SUBDIRS = libmodplug .
+SUBDIRS = .
 
 plugin_LTLIBRARIES = libgstmodplug.la
 
 libgstmodplug_la_SOURCES = gstmodplug.cc
 libgstmodplug_la_CXXFLAGS = $(GST_PLUGINS_BASE_CFLAGS) $(GST_CXXFLAGS)
-libgstmodplug_la_LIBADD = $(top_builddir)/gst/modplug/libmodplug/libmodplug.la $(GST_PLUGINS_BASE_LIBS) -lstdc++
+libgstmodplug_la_LIBADD = $(GST_PLUGINS_BASE_LIBS) -lstdc++ -lmodplug
 libgstmodplug_la_LDFLAGS = $(GST_PLUGIN_LDFLAGS)
 
 noinst_HEADERS = gstmodplug.h
@@ -547,8 +547,7 @@ CONFIG_HEADER = $(top_builddir)/config.h
 CONFIG_CLEAN_FILES =
 LTLIBRARIES = $(plugin_LTLIBRARIES)
 
-libgstmodplug_la_DEPENDENCIES = \
-	$(top_builddir)/gst/modplug/libmodplug/libmodplug.la
+libgstmodplug_la_DEPENDENCIES =
 am_libgstmodplug_la_OBJECTS = libgstmodplug_la-gstmodplug.lo
 libgstmodplug_la_OBJECTS = $(am_libgstmodplug_la_OBJECTS)
 
